planner/core/exhaust_physical_plans.go

```diff
@@ -529,100 +583,379 @@ func (p *LogicalJoin) constructInnerIndexScan(ds *DataSource, idx *model.IndexIn
+			return err
 		}
 	}
+	return nil
+}
 
-	// Filter out invalidFakeConds from res.RemainedConds.
-	for _, cond := range res.RemainedConds {
-		if !expression.Contains(invalidFakeConds, cond) {
-			remained = append(remained, cond)
+// String implements Stringer interface.
+func (cwc *ColWithCmpFuncManager) String() string {
+	buffer := bytes.NewBufferString("")
+	for i := range cwc.OpType {
+		buffer.WriteString(fmt.Sprintf("%v(%v, %v)", cwc.OpType[i], cwc.targetCol, cwc.opArg[i]))
+		if i < len(cwc.OpType)-1 {
+			buffer.WriteString(" ")
 		}
 	}
-
-	return res.Ranges, remained, keyOff2IdxOff
+	return buffer.String()
 }
 
-func (p *LogicalJoin) buildFakeEqCondsForIndexJoin(keys, idxCols []*expression.Column, colLengths []int,
-	innerFilters []expression.Expression) (accesses, eqConds, remained []expression.Expression, keyOff2IdxOff []int) {
-	// Check whether all join keys match one column from index.
-	keyOff2IdxOff = joinKeysMatchIndex(keys, idxCols, colLengths)
-	if keyOff2IdxOff == nil {
-		return nil, nil, nil, nil
+func (ijHelper *indexJoinBuildHelper) checkIndex(innerKeys []*expression.Column, idxCols []*expression.Column, colLens []int) bool {
+	tmpSchema := expression.NewSchema(innerKeys...)
+	ijHelper.curIdxOff2KeyOff = make([]int, len(idxCols))
+	ijHelper.curNotUsedIndexCols = make([]*expression.Column, 0, len(idxCols))
+	ijHelper.curNotUsedColLens = make([]int, 0, len(idxCols))
+	keyMatched := false
+	for i, idxCol := range idxCols {
+		ijHelper.curIdxOff2KeyOff[i] = tmpSchema.ColumnIndex(idxCol)
+		if ijHelper.curIdxOff2KeyOff[i] >= 0 {
+			keyMatched = true
+			continue
+		}
+		ijHelper.curNotUsedIndexCols = append(ijHelper.curNotUsedIndexCols, idxCol)
+		ijHelper.curNotUsedColLens = append(ijHelper.curNotUsedColLens, colLens[i])
 	}
+	return keyMatched
+}
 
-	usableKeys := make([]*expression.Column, 0, len(keys))
+// findUsefulEqAndInFilters analyzes the pushedDownConds held by inner child and split them to three parts.
+// usefulEqOrInFilters is the continuous eq/in conditions on current unused index columns.
+// uselessFilters is the conditions which cannot be used for building ranges.
+// remainingRangeCandidates is the other conditions for future use.
+func (ijHelper *indexJoinBuildHelper) findUsefulEqAndInFilters(innerPlan *DataSource) (usefulEqOrInFilters, uselessFilters, remainingRangeCandidates []expression.Expression) {
+	uselessFilters = make([]expression.Expression, 0, len(innerPlan.pushedDownConds))
+	remainingRangeCandidates = make([]expression.Expression, 0, len(innerPlan.pushedDownConds))
+	// This loop finds the possible filters which can be used to build ranges.
+	// If the filter contains index column covered by join keys, it will be useless since we always use join key to build range for that index column..
+	for _, innerFilter := range innerPlan.pushedDownConds {
+		affectedCols := expression.ExtractColumns(innerFilter)
+		if expression.ColumnSliceIsIntersect(affectedCols, ijHelper.curPossibleUsedKeys) {
```

Comment:
Seems that `curPossibleUsedKeys` is always empty here because we build them in `removeUselessEqAndInFunc`, which is after this.

Reply:
Removed this for loop.